staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1222,7 +1302,9 @@ func (c *cacheWatcher) add(event *watchCacheEvent, timer *time.Timer) bool {
 		// we simply terminate it.
 		klog.V(1).Infof("Forcing %v watcher close due to unresponsiveness: %v. len(c.input) = %v, len(c.result) = %v", c.objectType.String(), c.identifier, len(c.input), len(c.result))
 		metrics.TerminatedWatchersCounter.WithLabelValues(c.objectType.String()).Inc()
-		c.forget()
+
+		// TODO: determine if the watcher needs draining
+		c.forget(false)
```

Comment:
I started wondering if we will really need that option at all or whether the watcher will be able to determine it themselves.

I think this all boils down the answering the question when exactly the watcher should be draining.
The only situation when I believe it makes sense is basically the situation when:
- the watcher is requesting the "bookmark after sync"
- that bookmark is already in its channel

In other situations I think it doesn't bring any value. Does that make sense?
If so - then the watcher can effectively figure it out themselves and we won't be forced to bass any arguments anywhere.

WDYT?

Reply:
I think it is necessary, mainly because we need to cover the following cases:

1. client disconnect (watcher.Stop())
2. c.terminateAllWatchers()

So, even if we are in a draining mode, triggering any of the above conditions should stop the watcher.
Does it make sense?

Reply:
OK - I think it makes sense.
So I think you should now figure out the condition here and document it clearly in a comment.

Reply:
yes, ok.

Reply:
This now looks reasonable to me - resolving.